PSYKHANOOL/PSYKHANOOL.h

```diff
@@ -0,0 +1,664 @@
+      return typename PERSISTER::ImmutableOptionalType(nullptr);
+    }
+  }
+
+  void PushBack(const T& object) {
+    persister_.PersistPushBack(VectorStorage<T>::vector_.size(), object);
+    VectorStorage<T>::vector_.push_back(object);
+  }
+
+  void PopBack() {
+    if (!VectorStorage<T>::vector_.empty()) {
+      persister_.PersistPopBack(VectorStorage<T>::vector_.size());
+      VectorStorage<T>::vector_.pop_back();
+    } else {
+      throw CannotPopBackFromEmptyVectorException();
+    }
+  }
+
+ private:
+  PERSISTER& persister_;
+};
+
+template <typename T>
+class OrderedDictionaryStorage {
+ public:
+  using T_KEY = sfinae::ENTRY_KEY_TYPE<T>;
+  using T_MAP = std::map<T_KEY, T>;
+  T_MAP map_;
+  void DoInsert(const T& object) { map_[sfinae::GetKey(object)] = object; }
+};
+
+template <typename T, typename PERSISTER>
+class OrderedDictionaryAPI : protected OrderedDictionaryStorage<T> {
+ public:
+  typedef typename OrderedDictionaryStorage<T>::T_MAP T_MAP;
+  typedef typename OrderedDictionaryStorage<T>::T_KEY T_KEY;
+  explicit OrderedDictionaryAPI(PERSISTER* persister) : persister_(*persister) {}
+
+  bool Empty() const { return OrderedDictionaryStorage<T>::map_.empty(); }
+  size_t Size() const { return OrderedDictionaryStorage<T>::map_.size(); }
+
+  typename PERSISTER::MutableOptionalType operator[](sfinae::CF<T_KEY> key) {
+    auto iterator = OrderedDictionaryStorage<T>::map_.find(key);
+    if (iterator != OrderedDictionaryStorage<T>::map_.end()) {
+      return typename PERSISTER::MutableOptionalType(&iterator->second);
+    } else {
+      return typename PERSISTER::MutableOptionalType(nullptr);
+    }
+  }
+
+  typename PERSISTER::ImmutableOptionalType operator[](sfinae::CF<T_KEY> key) const {
+    const auto iterator = OrderedDictionaryStorage<T>::map_.find(key);
+    if (iterator != OrderedDictionaryStorage<T>::map_.end()) {
+      return &iterator->second;
+    } else {
+      return nullptr;
+    }
+  }
+
+  void Insert(const T& object) {
+    persister_.PersistInsert(object);
+    OrderedDictionaryStorage<T>::DoInsert(object);
+  }
+
+  void Erase(sfinae::CF<T_KEY> key) {
+    persister_.PersistErase(key);
+    OrderedDictionaryStorage<T>::map_.erase(key);
+  }
+
+  struct Iterator final {
+    using T_ITERATOR = typename T_MAP::const_iterator;
+    T_ITERATOR iterator;
+    explicit Iterator(T_ITERATOR iterator) : iterator(std::move(iterator)) {}
+    void operator++() { ++iterator; }
+    bool operator==(const Iterator& rhs) const { return iterator == rhs.iterator; }
+    bool operator!=(const Iterator& rhs) const { return !operator==(rhs); }
+    sfinae::CF<T_KEY> key() const { return iterator->first; }
+    const T& operator*() const { return iterator->second; }
+    const T* operator->() const { return &iterator->second; }
+  };
+
+  Iterator begin() const { return Iterator(OrderedDictionaryStorage<T>::map_.cbegin()); }
+  Iterator end() const { return Iterator(OrderedDictionaryStorage<T>::map_.cend()); }
```

Comment:
Perhaps, haven't looked into whether it would work with `for (const auto& e : container)` syntax.